### PR TITLE
feat: refactor fsdp for fp16/bf16 only

### PR DESCRIPTION
**Pull request by YimingZhou2002**
### Description

This PR refactors the FSDP precision handling logic to explicitly support fp16 and bf16 only, removing ambiguous or implicit precision paths and making the behavior more explicit and predictable.

In addition, the AMP configuration in example YAML files has been cleaned up and clarified.
The previously amp block has been replaced with a more explicit and semantically clear configuration using amp_autocast and amp_grad_scaler, improving readability and reducing potential misconfiguration. 

The code logic has been updated accordingly to handle the new configuration structure. Notably, backward compatibility is preserved: configurations using the legacy amp block (e.g., amp.precision, amp.enabled, amp.use_grad_scaler) are still supported during the transition period.

### Motivation

Supporting fp16/bf16 only in fsdp. Certain amp configurations are used in fsdp while amp is disabled could cause confusion among developers.

### How has this been tested?

A trial run on searchr1-fsdp reveal related parameters has been properly configured. A local pre-commit style check has been conducted.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: examples/reasoning/config/vqa/qwen2.5-vl-3b-grpo-fsdp.yaml [hunk 1]
```diff
@@ -231,11 +231,17 @@ actor:
...
+      enabled: False                    # if True, automatic mixed precision (AMP) will be used in FSDP/FSDP2 training.
+      precision: "bf16"                 # precision for AMP, options are ["fp16" , "bf16"]
+    
+    amp_grad_scaler: False               # if True, GradScaler will be used for AMP training to prevent underflow.
 
-    amp:
-      enabled: False                 # if True, automatic mixed precision (AMP) will be used in FSDP/FSDP2 training.
-      precision: "bf16"              # precision for AMP, options are ["fp16" , "bf16"]
-      use_grad_scaler: False         # if True, GradScaler will be used for AMP training to prevent underflow.
+    # amp:
```
Comment (qurakchin) [inline examples/reasoning/config/vqa/qwen2.5-vl-3b-grpo-fsdp.yaml]: delete these lines.

File: examples/reasoning/config/vqa/qwen2.5-vl-3b-grpo-fsdp.yaml [hunk 2]
```diff
@@ -231,11 +231,17 @@ actor:
       param_dtype: "bf16"
       reduce_dtype: "fp32"
       buffer_dtype: "fp32"
+    
+    amp_autocast:
```
Comment (qurakchin) [inline examples/reasoning/config/vqa/qwen2.5-vl-3b-grpo-fsdp.yaml]: add a comment that amp_autocast cannot use with mixed_precision simultaneously

File: examples/reasoning/config/vqa/qwen2.5-vl-3b-grpo-fsdp.yaml
Comment (qurakchin) [inline examples/reasoning/config/vqa/qwen2.5-vl-3b-grpo-fsdp.yaml]: add a comment that mixed_precision cannot use with amp_autocast simultaneously。

File: examples/reasoning/config/vqa/qwen2.5-vl-3b-grpo-fsdp.yaml (hunk 2 above)
Comment (qurakchin) [reply examples/reasoning/config/vqa/qwen2.5-vl-3b-grpo-fsdp.yaml]: apply to all the yaml if the yaml has the actor.amp section.

File: rlinf/utils/initialize.py [hunk 3]
```diff
@@ -66,14 +66,17 @@ def set_megatron_args(cfg):
...
+    precision_dtype = torch_dtype_from_precision(cfg.model.precision)
+    if precision_dtype == torch.float16:
+        args.params_dtype = "${torch.dtype:half}"
+    elif precision_dtype == torch.bfloat16:
+        args.params_dtype = "${torch.dtype:bfloat16}"
+    else:
+        assert False, (
+            "model.precision only available in fp16 or bf16 when using megatron"
+        )
+    args.fp16 = precision_dtype == torch.float16
```
Comment (qurakchin) [inline rlinf/utils/initialize.py]: assign the param args.fp16 only if the param is not exist.

File: rlinf/utils/initialize.py [hunk 4]
```diff
@@ -66,14 +66,17 @@ def set_megatron_args(cfg):
...
+    args.bf16 = precision_dtype == torch.bfloat16
```
Comment (qurakchin) [inline rlinf/utils/initialize.py]: same as the .fp16.

File: rlinf/config.py [hunk 5]
```diff
@@ -305,14 +307,42 @@ def validate_model_cfg_by_hf_config(cfg, hf_model_path):
...
     def validate_amp_cfg(config: DictConfig) -> DictConfig:
-        if "amp" not in config:
+        use_fsdp_mixed_precision = not (
+            mixed_precision_config.param_dtype is None
+            and mixed_precision_config.reduce_dtype is None
+            and mixed_precision_config.buffer_dtype is None
+        )
+        if "amp_autocast" in config or "amp_grad_scaler" in config:
+            assert "amp" not in config, (
+                "fsdp.amp_autocast and fsdp.amp_grad_scaler should not be enabled when fsdp.amp is used"
```
Comment (qurakchin) [inline rlinf/config.py]: should not be enabled -> should not used

**Review by qurakchin — CHANGES_REQUESTED** [review]

File: rlinf/hybrid_engines/fsdp/fsdp_model_manager.py [hunk 6]
```diff
@@ -259,9 +261,19 @@ def setup_model_and_optimizer(self) -> None:
...
+            for key in ["init_scale", "growth_interval"]:
+                value = self._cfg.fsdp_config.amp_grad_scaler.get(key, None)
+                if value is not None:
+                    kwargs[key] = value
+            self.grad_scaler = self.build_grad_scaler(
+                self._cfg.fsdp_config.amp_grad_scaler.get("enabled", False), **kwargs
+            )
+        else:
+            self.grad_scaler = self.build_grad_scaler(
+                self._cfg.fsdp_config.amp_grad_scaler
```
Comment (qurakchin) [inline rlinf/hybrid_engines/fsdp/fsdp_model_manager.py]: in your case, will self._cfg.fsdp_config.amp_grad_scaler always exist?

File: rlinf/config.py [hunk 7]
```diff
@@ -310,14 +312,58 @@ def validate_model_cfg_by_hf_config(cfg, hf_model_path):
...
+        param_dtype = mixed_precision_config.get("param_dtype", None)
+        reduce_dtype = mixed_precision_config.get("reduce_dtype", None)
+        buffer_dtype = mixed_precision_config.get("buffer_dtype", None)
+
+        use_fsdp_mixed_precision = not (
+            param_dtype is None and reduce_dtype is None and buffer_dtype is None
+        )
+        if "amp_autocast" in config or "amp_grad_scaler" in config:
+            assert "amp" not in config, (
+                "fsdp.amp_autocast and fsdp.amp_grad_scaler should not be used when fsdp.amp is used"
```
Comment (qurakchin) [inline rlinf/config.py]: -> "fsdp.amp should not be used when fsdp.amp_autocast or fsdp.amp_grad_scaler is used"

File: rlinf/config.py [hunk 8]
```diff
@@ -310,14 +312,58 @@ def validate_model_cfg_by_hf_config(cfg, hf_model_path):
...
+            )
             config.amp = {}
-        config.amp.enabled = config.amp.get("enabled", False)
-        config.amp.precision = config.amp.get("precision", "bf16")
-        assert config.amp.precision in ["fp16", "bf16", "fp32"], (
-            "fsdp.amp.precision must be one of ['fp16', 'bf16', 'fp32']"
+            config.amp_autocast.enabled = config.amp_autocast.get("enabled", False)
```
Comment (qurakchin) [inline rlinf/config.py]: use style like the following process of config.amp_grad_scaler

File: rlinf/config.py [hunk 9]
```diff
@@ -310,14 +312,58 @@ def validate_model_cfg_by_hf_config(cfg, hf_model_path):
...
+        buffer_dtype = mixed_precision_config.get("buffer_dtype", None)
+
+        use_fsdp_mixed_precision = not (
+            param_dtype is None and reduce_dtype is None and buffer_dtype is None
+        )
+        if "amp_autocast" in config or "amp_grad_scaler" in config:
+            assert "amp" not in config, (
+                "fsdp.amp_autocast and fsdp.amp_grad_scaler should not be used when fsdp.amp is used"
+            )
             config.amp = {}
```
Comment (qurakchin) [inline rlinf/config.py]: delete this line

File: rlinf/config.py [hunk 10]
```diff
@@ -310,14 +312,58 @@ def validate_model_cfg_by_hf_config(cfg, hf_model_path):
...
+            if "amp" not in config:
+                config.amp = {}
+            config.amp_autocast.enabled = config.amp.get("enabled", False)
+            if config.amp_autocast.enabled is False:
+                if "precision" in config.amp:
+                    logging.warning(
+                        "fsdp_config.amp.precision will be deprecated under amp disabled condition, use fsdp_config.amp_autocast.precision instead"
+                    )
+                if "use_grad_scaler" in config.amp:
+                    logging.warning(
```
Comment (qurakchin) [inline rlinf/config.py]: amp.use_grad_scaler is independent with amp.enable

File: tests/e2e_tests/embodied/frankasim_ppo_mlp.yaml [hunk 11]
```diff
@@ -169,11 +169,13 @@ actor:
       reduce_dtype: ${actor.model.precision}
       buffer_dtype: ${actor.model.precision}
 
-
-    amp:
-      enabled: false
-      precision: bf16
-      use_grad_scaler: false
+
```
Comment (qurakchin) [inline tests/e2e_tests/embodied/frankasim_ppo_mlp.yaml]: is this case tested reward line?

Comment (YimingZhou2002) [reply tests/e2e_tests/embodied/frankasim_ppo_mlp.yaml]: These fixed lines within embodied configurations has been roll-backed to align with previous releases.

File: rlinf/config.py (hunk 8 above)
Comment (YimingZhou2002) [reply rlinf/config.py]: amp config backward compatablity process has been improved to align with grad scaler like dict  var assignment

**Review by qurakchin — APPROVED** [review]

File: tests/e2e_tests/embodied/frankasim_ppo_mlp.yaml [hunk 12]
```diff
@@ -169,11 +169,11 @@ actor:
       reduce_dtype: ${actor.model.precision}
       buffer_dtype: ${actor.model.precision}
 
-
+ 
     amp:
       enabled: false
       precision: bf16
-      use_grad_scaler: false
+      use_grad_scaler: false
```
Comment (qurakchin) [inline tests/e2e_tests/embodied/frankasim_ppo_mlp.yaml]: revert the space and black line change

File: docs/source-en/rst_source/tutorials/user/yaml.rst [hunk 13]
```diff
@@ -546,11 +546,13 @@ actor
...
-      amp:
-        enabled: False
-        precision: "bf16"
-        use_grad_scaler: False
+     
+      amp_autocast:                                
+        enabled: False                    
+        precision: "bf16"                 
+      
+      amp_grad_scaler:
```
Comment (thereAreDemonsNearby) [inline docs/source-en/rst_source/tutorials/user/yaml.rst]: why is this named 'amp_grad_scaler' ? AFAIK, grad scaler is needed when you train with fp16, no matter torch amp is enabled or not

Comment (YimingZhou2002) [reply docs/source-en/rst_source/tutorials/user/yaml.rst]: As grad scaler value come into effect regardless of the status of amp.enable variable, amp_grad_scaler is stripped from the main amp (amp_autocast)
structure to clearify the relationship between amp autocast and the grad scaler.

Comment (thereAreDemonsNearby) [reply docs/source-en/rst_source/tutorials/user/yaml.rst]: What I mean is, is this name appropriate?  Would simply 'grad_scaler' be better?

Comment (YimingZhou2002) [reply docs/source-en/rst_source/tutorials/user/yaml.rst]: amp_grad_scaler naming has been adjusted accordingly as requested. amp_grad_scaler ->amp_grad_scaler

**Review by thereAreDemonsNearby — APPROVED** [review]

File: rlinf/hybrid_engines/fsdp/fsdp_model_manager.py [hunk 14]
```diff
@@ -104,11 +104,13 @@ def _create_amp_context(self) -> ContextManager:
         """
         from contextlib import nullcontext
 
-        if not self._cfg.fsdp_config.amp.enabled:
+        if not self._cfg.fsdp_config.amp_autocast.enabled:
```
Comment (andylin-hao) [inline rlinf/hybrid_engines/fsdp/fsdp_model_manager.py:107]: Doesn't this break all embodied config?

**Review by andylin-hao — APPROVED** [review]

File: rlinf/hybrid_engines/fsdp/fsdp_model_manager.py [hunk 15]
```diff
@@ -263,8 +265,18 @@ def setup_model_and_optimizer(self) -> None:
         self.lr_scheduler = self.build_lr_scheduler(
             optimizer=self.optimizer, optim_config=self._cfg.optim
         )
+
+        assert self._cfg.fsdp_config.get("grad_scaler") is not None, (
```
Comment (andylin-hao) [inline rlinf/hybrid_engines/fsdp/fsdp_model_manager.py:269]: Same here. What about embodied configs?

File: rlinf/config.py [hunk 16]
```diff
@@ -322,39 +322,24 @@ def validate_amp_cfg(config: DictConfig) -> DictConfig:
...
+        grad_scaler = config.get("grad_scaler", {})
+        config.grad_scaler = {
+            "enabled": grad_scaler.get("enabled", False),
+            "init_scale": grad_scaler.get("init_scale", None),
+            "growth_interval": grad_scaler.get("growth_interval", None),
+        }
+
+        if "amp" in config:
+            logging.warning(
+                "fsdp.amp is no longer supported, use fsdp.amp_autocast and fsdp.grad_scaler instead"
```
Comment (qurakchin) [inline rlinf/config.py]: -> fsdp_config.amp

File: rlinf/config.py [hunk 17]
```diff
@@ -409,16 +394,6 @@ def validate_amp_cfg(config: DictConfig) -> DictConfig:
             "fsdp_config.mixed_precision is required in FSDP actor configuration."
         )
 
-        mixed_precision_config = cfg.fsdp_config.mixed_precision
```
Comment (qurakchin) [inline rlinf/config.py:368]: why delete these lines

File: rlinf/config.py [hunk 18]
```diff
@@ -310,14 +312,41 @@ def validate_model_cfg_by_hf_config(cfg, hf_model_path):
...
-            "fsdp.amp.precision must be one of ['fp16', 'bf16', 'fp32']"
-        )
-        config.amp.use_grad_scaler = config.amp.get("use_grad_scaler", False)
+        """Validate AMP configuration and ensure mutual exclusivity with FSDP mixed_precision."""
+
+        param_dtype = config.mixed_precision.param_dtype
+        reduce_dtype = config.mixed_precision.reduce_dtype
+        buffer_dtype = config.mixed_precision.buffer_dtype
+
+        use_fsdp_mixed_precision = not (
```
Comment (qurakchin) [inline rlinf/config.py]: when all 3 dtypes are fp32, use_fsdp_mixed_precision is True

File: examples/reasoning/config/math/qwen2.5-1.5b-grpo-fsdp-pipeline.yaml [hunk 19]
```diff
@@ -260,15 +260,17 @@ actor:
...
       param_dtype: "fp16"
       reduce_dtype: "fp16"
       buffer_dtype: "fp32"
 
-    amp:
-      enabled: True                 # if True, automatic mixed precision (AMP) will be used in FSDP/FSDP2 training.
-      precision: "fp16"              # precision for AMP, options are ["fp16" , "bf16"]
-      use_grad_scaler: True          # if True, GradScaler will be used for AMP training to prevent underflow.
+    amp_autocast:                       # Do not enable `amp_autocast` with FSDP `mixed_precision` enabled, using both will result in conflict.  
+      enabled: True                    # if True, automatic mixed precision (AMP) will be used in FSDP/FSDP2 training.
```
Comment (qurakchin) [inline examples/reasoning/config/math/qwen2.5-1.5b-grpo-fsdp-pipeline.yaml]: use False

**Review by guozhen1997 — APPROVED** [review]

**Review by andylin-hao — APPROVED** [review]
